src/python/pants/backend/python/rules/setup_py_util.py

```diff
@@ -93,52 +96,117 @@ def _write_repr(o, indent=False, level=0):
+  init_py_by_path: Dict[str, bytes] = {ipc.path: ipc.content for ipc in init_py_contents}
+  for pkg in packages:
+    path = os.path.join(pkg.replace('.', os.path.sep), '__init__.py')
+    if (path in init_py_by_path and
+        declares_pkg_resources_namespace_package(init_py_by_path[path].decode())):
+      namespace_packages.add(pkg)
 
   return (tuple(sorted(packages)),
           tuple(sorted(namespace_packages)),
           tuple((pkg, tuple(sorted(files))) for pkg, files in package_data.items()))
+
+
+def declares_pkg_resources_namespace_package(python_src: str) -> bool:
+  """Given .py file contents, determine if it declares a pkg_resources-style namespace package.
+
+  Detects pkg_resources-style namespaces. See here for details:
+  https://packaging.python.org/guides/packaging-namespace-packages/.
+
+  Note: Accepted namespace package decls are valid Python syntax in all Python versions,
+  so this code can, e.g., detect namespace packages in Python 2 code while running on Python 3.
+  """
+  import ast
+
+  def is_name(node: ast.AST, name: str) -> bool:
+    return isinstance(node, ast.Name) and node.id == name
+
+  def is_call_to(node: ast.AST, func_name: str) -> bool:
+    if not isinstance(node, ast.Call):
+      return False
+    func = node.func
+    return (isinstance(func, ast.Attribute) and func.attr == func_name) or is_name(func, func_name)
+
+  def has_args(call_node: ast.Call, required_arg_ids: Tuple[str, ...]) -> bool:
+    args = call_node.args
+    if len(args) != len(required_arg_ids):
+      return False
+    actual_arg_ids = tuple(arg.id for arg in args if isinstance(arg, ast.Name))
+    return actual_arg_ids == required_arg_ids
+
+  try:
+    python_src_ast = ast.parse(python_src)
+  except SyntaxError:
+    # The namespace package incantations we check for are valid code in all Python versions.
+    # So if the code isn't parseable we know it isn't a valid namespace package.
+    return False
+
+  # Note that these checks are slightly heuristic. It is possible to construct adversarial code
+  # that would defeat them. But the only consequence would be an incorrect namespace_packages list
+  # in setup.py, and we're assuming our users aren't trying to shoot themselves in the foot.
+  for ast_node in ast.walk(python_src_ast):
+    # pkg_resources-style namespace, e.g.,
+    #   __import__('pkg_resources').declare_namespace(__name__).
+    if (is_call_to(ast_node, 'declare_namespace') and
+        has_args(cast(ast.Call, ast_node), ('__name__',))):
+      return True
+  return False
+
+
+def is_python2(compatibilities: Iterable[Optional[List[str]]], python_setup: PythonSetup) -> bool:
+  """Checks if we should assume python2 code."""
+
+  reqs = [Requirement.parse(constraint)
+          for compatibility in compatibilities
+          for constraint in python_setup.compatibility_or_constraints(compatibility)]
+  for req in reqs:
+    if req.specifier.contains('2.7'):  # type: ignore
```

Comment:
This is probably fine in practice but its also definitely not right. It might be worth a comment / TODO pointing out !=2.7.* is a completely valid specifier that this trips up on. I can't see anyone writing this down though so I'm not too worried.

Reply:
Good point. It's actually fairly easy to just check for 2.7.0 ... 2.7.17, so I did that.